### PR TITLE
WIP: referenced file cache

### LINKED ISSUE #13: Improve resolving references, cache loaded files in current context
To avoid multiple requests to the same file, loaded files can be cached in current context.

### LINKED ISSUE #63: Add option to inline only external references
Currently resolving references resolves all references unless they are recursive.

There should be more options to support cases where only external references should be inlined.
See also:
https://stoplight.io/blog/keeping-openapi-dry-and-portable/

### PR DESCRIPTION
**Pull request by cebe**
- fixes #13 
- partially fixes #63

### CODE REVIEW

**Comment by araines** [comment]
@cebe is there anything I can do to help with this one?  Your tool is great, but we cannot use it with our project as we make heavy usage of references and the currently released version gets stuck in loops.  It looks to me like this fix should actually resolve that issue.

**Comment by cebe** [comment]
These changes are currently creating a bug in one of my project so it is not complete. I can try to provide a test case for it so you can check it out. Thanks for offering your help.

**Comment by kohlerdominik** [comment]
Hi @cebe 
Is there an ETA for this feature?

**Comment by cebe** [comment]
No ETA, but I try to get it done soon ;-)

**Comment by stevethitchman** [comment]
Hi @cebe did you manage to implement a test case for the bug you were having on one of your projects, it looks like you may have? I may be able to put some time into this if so as it's recently become an issue on a project and from a quick test these changes seem to pretty much resolve it

File: src/spec/Reference.php [hunk 1]
```diff
@@ -205,35 +196,39 @@ public function resolve(ReferenceContext $context = null)
...
             // resolve in external document
             $file = $context->resolveRelativeUri($jsonReference->getDocumentUri());
-            // TODO could be a good idea to cache loaded files in current context to avoid loading the same files over and over again
-            $referencedDocument = $this->fetchReferencedFile($file);
-            $referencedData = $jsonReference->getJsonPointer()->evaluate($referencedDocument);
-
-            if ($referencedData === null) {
-                return null;
+            try {
+                // TODO cache here?
```
Comment (cebe) [inline src/spec/Reference.php:200]:
```suggestion
```

File: src/Reader.php [hunk 2]
```diff
@@ -57,9 +58,11 @@ public static function readFromYaml(string $yaml, string $baseType = OpenApi::cl
      * @param string $baseType the base Type to instantiate. This must be an instance of [[SpecObjectInterface]].
      * The default is [[OpenApi]] which is the base type of a OpenAPI specification file.
      * You may choose a different type if you instantiate objects from sub sections of a specification.
-     * @param bool $resolveReferences whether to automatically resolve references in the specification.
+     * @param bool|string $resolveReferences whether to automatically resolve references in the specification.
      * If `true`, all [[Reference]] objects will be replaced with their referenced spec objects by calling
      * [[SpecObjectInterface::resolveReferences()]].
+     * Since version 1.5.0 this can be a string indicating the reference resolving mode:
+     * - TODO inline vs all
```
Comment (cebe) [inline src/Reader.php:65]:
```suggestion
     * - `inline` only resolve references to external files.
     * - `all` resolve all references exceot recursive references.
```

File: src/Reader.php [hunk 3]
```diff
@@ -90,9 +98,11 @@ public static function readFromJsonFile(string $fileName, string $baseType = Ope
...
```
Comment (cebe) [inline src/Reader.php:105]:
```suggestion
     * - `inline` only resolve references to external files.
     * - `all` resolve all references exceot recursive references.
```

**Comment by cebe** [comment]
@stevethitchman thanks for offering your help. It turned out to be an issue in the project code.

**Pull request merged**